src/main/java/com/google/googleidentity/oauth2/endpoint/TokenEndpoint.java

```
@@ -0,0 +1,139 @@
+import com.google.googleidentity.oauth2.exception.OAuth2Exception;
+import com.google.googleidentity.oauth2.exception.OAuth2ExceptionHandler;
+import com.google.googleidentity.oauth2.request.OAuth2Request;
+import com.google.googleidentity.oauth2.util.OAuth2ParameterNames;
+import com.google.googleidentity.oauth2.util.OAuth2Utils;
+import com.google.googleidentity.oauth2.validator.TokenEndpointRequestValidator;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.util.logging.Logger;
+
+/**
+ * Token exchange endpoint in OAuth2 Server
+ */
+@Singleton
+public class TokenEndpoint extends HttpServlet {
+
+    private static final long serialVersionUID = 5L;
+
+    private static final Logger log = Logger.getLogger("TokenEndpoint");
+
+    private final ClientDetailsService clientDetailsService;
+
+
+    @Inject
+    public TokenEndpoint(ClientDetailsService clientDetailsService) {
+        this.clientDetailsService = clientDetailsService;
+    }
+
+    protected void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException, UnsupportedOperationException {
+
+        OAuth2Exception exception =
+                new InvalidRequestException(
+                        InvalidRequestException.ErrorCode.UNSUPPORTED_REQUEST_METHOD);
+        log.info("Token endpoint does not support GET request." );
+        response.setStatus(exception.getHttpCode());
+        response.setContentType("application/json;charset=UTF-8");
+        response.getWriter().println(
+                OAuth2ExceptionHandler.getResponseBody(exception).toJSONString());
+        response.getWriter().flush();
```

Can you just throw InvalidRequestException and rely on OAuth2ExceptionHandler to format the response?

Same with doPost() validation exception below.